src/main/java/at/hannibal2/skyhanni/features/inventory/caketracker/CakeTracker.kt

```diff
@@ -0,0 +1,366 @@
+    private var currentYear = 0
+
+    private var inCakeBag = false
+    private var inCakeInventory = false
+    private var timeOpenedCakeInventory = SimpleTimeMark.farPast()
+    private var inAuctionHouse = false
+    private var auctionCakesCache = mapOf<Int, Boolean>()
+    private var searchingForCakes = false
+    private var knownCakesInCurrentInventory = listOf<Int>()
+
+    /**
+     * REGEX-TEST: §cNew Year Cake (Year 360)
+     * REGEX-TEST: §cNew Year Cake (Year 1,000)
+     * REGEX-TEST: §f§f§cNew Year Cake (Year 330)
+     */
+    private val cakeNamePattern by patternGroup.pattern(
+        "cake.name",
+        "(?:§f§f)?§cNew Year Cake \\(Year (?<year>[\\d,]+)\\)",
+    )
+
+    /**
+     * REGEX-TEST: Ender Chest (2/9)
+     * REGEX-TEST: Jumbo Backpack (Slot #6)
+     * REGEX-TEST: New Year Cake Bag
+     */
+    private val cakeContainerPattern by patternGroup.pattern(
+        "cake.container",
+        "Ender Chest \\(\\d{1,2}/\\d{1,2}\\)|.*Backpack(?:§r)? \\(Slot #\\d{1,2}\\)|New Year Cake Bag",
+    )
+
+    /**
+     * REGEX-TEST: New Year Cake Bag
+     */
+    private val cakeBagPattern by patternGroup.pattern(
+        "cake.bag",
+        "New Year Cake Bag",
+    )
+
+    /**
+     * REGEX-TEST: Auctions Browser
+     * REGEX-TEST: Auctions: "Test"
+     */
+    private val auctionBrowserPattern by patternGroup.pattern(
+        "auction.search",
+        "Auctions Browser|Auctions: \".*",
+    )
+
+    /**
+     * REGEX-TEST: Auctions: "New Year C
+     */
+    private val auctionCakeSearchPattern by patternGroup.pattern(
+        "auction.cakesearch",
+        "^Auctions: \"New Year C.*$",
+    )
+
+    private fun addCake(cakeYear: Int) {
+        val storage = storage ?: return
+        if (cakeYear !in storage.ownedCakes) storage.ownedCakes.add(cakeYear)
+        recalculateMissingCakes()
+    }
+
+    private fun removeCake(cakeYear: Int) {
+        val storage = storage ?: return
+        if (cakeYear in storage.neededCakes) storage.ownedCakes.remove(cakeYear)
+        recalculateMissingCakes()
+    }
+
+    private fun isEnabled() = LorenzUtils.inSkyBlock && config.enabled
+
+    @HandleEvent
+    fun onCommandRegistration(event: CommandRegistrationEvent) {
+        event.register("shresetcaketracker") {
+            description = "Resets the New Year Cake Tracker"
+            category = CommandCategory.USERS_RESET
+            callback {
+                val storage = storage ?: return@callback
+                storage.ownedCakes.clear()
+                recalculateMissingCakes()
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onBackgroundDraw(event: GuiRenderEvent.ChestGuiOverlayRenderEvent) {
+        if (!isEnabled()) return
+        if (inCakeBag || (inAuctionHouse && (auctionCakesCache.any() || searchingForCakes))) {
+            config.cakeTrackerPosition.renderRenderables(
+                drawDisplay(storage ?: return),
```

Comment:
cache the display and only update it when necessary, instead of having to create it every frame